service/chain_service.go

```diff
@@ -97,6 +97,11 @@ func (cs *chainService) Add(bs domains.BlockHeaderSource) (*domains.BlockHeader,
 		return nil, err
 	}
 
+	currentTip, errTip := cs.Headers.GetTip()
```

Comment:
As we are calculating hash in the first line of this method, why not just check if there isn't already header with the same hash stored in database. If there is, then for example return an error like: `header already in database`